File: patterns/snake.go
package patterns

import (
	"github.com/anderstorpsfestivalen/slisko/pkg/chassi"
)

type Snake struct {
	snakeFrame int64
}

func (p *Snake) Render(info RenderInfo, c *chassi.Chassi) {
	if p.snakeFrame > int64(len(c.LEDs)) {
		p.snakeFrame = 0
	} else {
		p.snakeFrame = p.snakeFrame + 1
	}

	for m, port := range c.LEDs {
		if m == int(p.snakeFrame) {
			port.SetClamped(0.5, 0.5, 0.5)
		} else {
			port.SetClamped(0.0, 0.0, 0.0)
		}
	}
}

func (p *Snake) Info() PatternInfo {
	return PatternInfo{
		Name:     "snake",
		Category: "global",
	}
}

func (p *Snake) Bootstrap(c *chassi.Chassi) {

}

File: engine/src/patterns/snake.rs
use crate::chassi::Chassi;
use crate::pattern::{BootstrapCtx, Pattern, PatternInfo, RenderInfo};
use crate::utils;

/// A single lit pixel sweeping the strand.
#[derive(Default)]
pub struct Snake;

impl Pattern for Snake {
    fn render(&mut self, info: &RenderInfo, c: &mut Chassi) {
        let n = c.leds.len();
        let lit = (utils::sin(info.secs, 1.0) * n as f32) as usize;
        for (index, pixel) in c.leds.iter_mut().enumerate() {
            if index == lit {
                pixel.set_clamped(1.0, 1.0, 0.5);
            } else {
                pixel.set_clamped(0.0, 0.0, 0.0);
            }
        }
    }

    fn info(&self) -> PatternInfo {
        PatternInfo {
            name: "snake",
            category: "global",
        }
    }

    fn bootstrap(&mut self, _c: &Chassi, _ctx: &mut BootstrapCtx) {}
}
